# Translational Neuromodeling: Exercise 5 - Variational Bayes
Spring Semester 2025

---

# 5.1: VB update equations

## 5.1 a

1. **Noise distribution**

   $$
   p(\varepsilon)=\sqrt{\frac{\tau}{2\pi}}\,
   \exp\!\left(-\frac{\tau}{2}\varepsilon^{2}\right),
   \qquad
   \log p(\varepsilon)=\tfrac12\log\tau-\tfrac12\log(2\pi)-\tfrac{\tau}{2}\varepsilon^{2}.
   $$

2. **Relate observations to noise**  

   With $y_n=\mu+\varepsilon_n$, each $\varepsilon_n = y_n-\mu$.

3. **Independent observations**

   $$
   \log p(\mathbf y\mid\mu,\tau)
   =\sum_{n=1}^{N}\log p(y_n\mid\mu,\tau)
   =\sum_{n=1}^{N}\left[\frac12\log\tau-\frac12\log(2\pi)-\frac{\tau}{2}(y_n-\mu)^2\right].
   $$

4. **Collect terms**

   $$
   \log p(\mathbf y\mid\mu,\tau)
   =\frac{N}{2}\log\tau
   -\frac{\tau}{2}\sum_{n=1}^{N}(y_n-\mu)^2
   -\frac{N}{2}\log(2\pi)
   $$

## 5.1 b

1. **Prior over the parameters** (Eq (4))

   $$
   p(\theta)
     = \sqrt{\frac{\lambda_0 \tau}{2\pi}}
       \exp\!\Bigl[-\tfrac{\lambda_0\tau}{2}\,(\mu-\mu_0)^2\Bigr]\;
       \frac{b_0^{\,a_0}\,\tau^{a_0-1}}{\Gamma(a_0)}\,
       \exp(-b_0\tau).
   $$

   Taking the logarithm,

   $$
   \log p(\theta)=
     \tfrac12\log\lambda_0
     +\tfrac12\log\tau
     -\tfrac12\log(2\pi)
     -\tfrac{\lambda_0\tau}{2}(\mu-\mu_0)^2
     +a_0\log b_0
     +(a_0-1)\log\tau
     -\log\Gamma(a_0)
     -b_0\tau.
   $$

2. **Add the log‑likelihood** (from a)

   $$
   \log p(\mathbf y\mid\mu,\tau)
   =\frac{N}{2}\log\tau
    -\frac{\tau}{2}\sum_{n=1}^{N}(y_n-\mu)^2
    -\frac{N}{2}\log(2\pi).
   $$

3. **Combine and collect terms**

   $$
   %
   \begin{aligned}
   \log p(\mathbf y,\theta)
   &=\left(\frac{N}{2}+a_0-\frac12\right)\log\tau
     -\frac{\tau}{2}\sum_{n=1}^{N}(y_n-\mu)^2
     -\frac{\lambda_0\tau}{2}(\mu-\mu_0)^2
     -b_0\tau \\[4pt]
   &\quad
     -\frac{N+1}{2}\log(2\pi)
     +\tfrac12\log\lambda_0
     +a_0\log b_0
     -\log\Gamma(a_0).
   \end{aligned}
   $$

## 5.1 c

Using the mean‑field rule,

$$
\log q^{*}(\mu)
\,=\, \mathbb{E}_{q(\tau)}\!\bigl[\log p(\mathbf y, \mu, \tau)\bigr] \;+\; \text{const}.
$$

From part (b) the $\mu$‑dependent part of the log‑joint is  

$$
-\frac{\tau}{2}\sum_{n=1}^{N}(y_n-\mu)^2
\;-\;
\frac{\lambda_0 \tau}{2}(\mu-\mu_0)^2.
$$

Taking the expectation over $\tau$ (i.e.\ replacing $\tau$ by its mean $\mathbb{E}_{\tau}[\tau]$) yields  

$$
%
\log q^{*}(\mu)
\,=\,
-\frac{1}{2}\,\mathbb{E}_{\tau}[\tau]
\Bigl(
\sum_{n=1}^{N}(y_n-\mu)^2
\;+\;
\lambda_0(\mu-\mu_0)^2
\Bigr)
\;+\;
\text{const}
$$

## 5.1 d

#### 1. Expand the two squared terms  


\begin{aligned}
\sum_{n=1}^{N}(y_n-\mu)^2
&= \sum_{n=1}^{N}\bigl(y_n^2 - 2\mu y_n + \mu^2\bigr)
  = \sum_{n=1}^{N}y_n^2
    -2\mu\sum_{n=1}^{N}y_n
    +N\mu^2, \\[4pt]
\lambda_0(\mu-\mu_0)^2
&= \lambda_0\bigl(\mu^2 -2\mu\mu_0 + \mu_0^2\bigr).
\end{aligned}


#### 2. Collect the $\mu$‑dependent pieces  

Group the coefficients of $\mu^2$ and $\mu$:


\begin{aligned}
\sum_{n=1}^{N}(y_n-\mu)^2
  +\lambda_0(\mu-\mu_0)^2
&= (N+\lambda_0)\mu^2
   -2\mu\bigl(N\bar y + \lambda_0\mu_0\bigr)
   +\text{const},
\end{aligned}

where $\displaystyle \bar y \;=\;\frac1N\sum_{n=1}^{N}y_n$.

#### 3. Insert back into $\log q^{*}(\mu)$  

$$
\boxed{%
\log q^{*}(\mu)
= -\frac{\bar\tau\,(N+\lambda_0)}{2}\,\mu^{2}
  \;+\;\bar\tau\,\mu\bigl(\lambda_0\mu_0 + N\bar y\bigr)
  \;+\;\text{const}}
$$

which is exactly equation (7) of the exercise.

## 5.1 e

#### 1. Arrange in the canonical quadratic form  

Factor out $\tfrac{\bar\tau(N+\lambda_0)}{2}$:

$$
\log q^{*}(\mu)
= -\frac{\bar\tau (N+\lambda_0)}{2}
  \Bigl[
      \mu^{2}
      - 2\mu\,
        \frac{\lambda_0\mu_0 + N\bar y}{N+\lambda_0}
  \Bigr]
  + \text{const}.
$$

---

#### 2. Compare with the log‑Gaussian template  

For a Gaussian $ \mathcal N(\mu \mid m, s^{2}) $  

$$
\log \mathcal N(\mu \mid m,s^{2})
= -\frac{1}{2s^{2}}\bigl(\mu-m\bigr)^{2} + \text{const}.
$$

Matching coefficients gives

$$
\boxed{%
s^{2} \;=\; \frac{1}{\bar\tau\,(N+\lambda_0)}
}
\qquad\text{and}\qquad
\boxed{%
m \;=\; \frac{\lambda_0\mu_0 + N\bar y}{\lambda_0 + N}
}.
$$

Hence $q^{*}(\mu)=\mathcal N\!\bigl(\mu\,\big|\,m,\;s^{2}\bigr)$ with the variance $s^{2}$ and mean $m$ exactly as in equations (8)–(9).

## 5.1 f

Under the mean‑field assumption $q(\theta)=q(\mu)\,q(\tau)$ we use
$$
\log q^{\!*}(\tau)
  \;=\;
  \mathbb{E}_{\mu}\!\bigl[\log p(\mathbf y, \mu, \tau)\bigr]
  \;+\;\text{const}.
$$

From part (b) the $\tau$‑dependent part of the log‑joint is  
$$
\left(\frac{N}{2}+a_0-\frac12\right)\!\log\tau
\;-\;\frac{\tau}{2}\sum_{n=1}^{N}(y_n-\mu)^2
\;-\;\frac{\lambda_0\tau}{2}(\mu-\mu_0)^2
\;-\;b_0\tau.
$$

Taking the expectation over $\mu$ (only the quadratic terms involve $\mu$) gives  
$$
\boxed{
\log q^{\!*}(\tau)
= -\frac{\tau}{2}\,
  \mathbb{E}_{\mu}\!\Bigl[\,
      \sum_{n=1}^{N}(y_n-\mu)^2
      + \lambda_0(\mu-\mu_0)^2
    \Bigr]
  \;-\; b_0\tau
  \;+\;
  \Bigl(
    a_0 + \frac{N+1}{2} - 1
  \Bigr)\log\tau
  \;+\;\text{const}}
$$

which matches equation (10).

## 5.1 g
$\displaystyle \mathbb{E}_{\mu}\!\bigl\{(x-\mu)^{2}\bigr\} = (x-m)^{2}+s^{2}$  

Recall the result from f)

$$
\log q^{*}(\tau)
 = -\frac{\tau}{2}\,
   \mathbb{E}_{\mu}\!\Bigl[
     \sum_{n=1}^{N}(y_n-\mu)^{2}
     + \lambda_0(\mu-\mu_0)^{2}
   \Bigr]
   - b_0\tau
   +\Bigl(a_0+\frac{N+1}{2}-1\Bigr)\log\tau
   + \text{const}.
$$

---

#### Apply the expectation identity term‑by‑term  

*For each data point*  
$$
\mathbb{E}_{\mu}\!\bigl[(y_n-\mu)^{2}\bigr]
  = (y_n-m)^{2}+s^{2}.
$$

*For the prior quadratic term*  
$$
\mathbb{E}_{\mu}\!\bigl[(\mu-\mu_0)^{2}\bigr]
  = (m-\mu_0)^{2}+s^{2}.
$$

Insert these results:

$$
\begin{aligned}
\mathbb{E}_{\mu}\!\Bigl[
  \sum_{n=1}^{N}(y_n-\mu)^{2}
  + \lambda_0(\mu-\mu_0)^{2}
\Bigr]
&= \sum_{n=1}^{N}(y_n-m)^{2}
   + \lambda_0(m-\mu_0)^{2}
   + (N+\lambda_0)s^{2}.
\end{aligned}
$$

---

#### Substitute back into $\log q^{*}(\tau)$  

$$
\boxed{
\log q^{*}(\tau)
 = -\frac{\tau}{2}\Bigl(
      \sum_{n=1}^{N}(y_n-m)^{2}
      + \lambda_0(\mu_0-m)^{2}
      + (N+\lambda_0)s^{2}
    \Bigr)
   - b_0\tau
   +\Bigl(a_0+\frac{N+1}{2}-1\Bigr)\log\tau
   + \text{const}}
$$

which is equation (11).

## 5.1 h

Equation (11) (from part (g)) is  

$$
\log q^{*}(\tau)
= -\frac{\tau}{2}
   \Bigl[
     \sum_{n=1}^{N}(y_n-m)^{2}
     + \lambda_0(\mu_0-m)^{2}
     + (N+\lambda_0)s^{2}
   \Bigr]
   \;-\; b_0\tau
   +\Bigl(a_0+\frac{N+1}{2}-1\Bigr)\log\tau
   + \text{const}.
$$

---

#### Compare with the log‑Gamma template  

For a Gamma distribution  
$$
\operatorname{Gam}(\tau\mid a,b)
  \;=\;
  \frac{b^{a}\,\tau^{a-1}}{\Gamma(a)}\,e^{-b\tau},
$$
we have  
$$
\log \operatorname{Gam}(\tau\mid a,b)
  \;=\;
  (a-1)\log\tau \;-\; b\tau \;+\; \text{const}.
$$

---

#### Match coefficients  

* **Coefficient of $\log\tau$**

  $$
  a-1
  \;=\;
  a_0 + \frac{N+1}{2} - 1
  \quad\Longrightarrow\quad
  \boxed{\,a = a_0 + \dfrac{N+1}{2}\,}.
  $$

* **Coefficient of $\tau$**

  $$
  b
  \;=\;
  b_0 + \tfrac12
      \Bigl[
        \sum_{n=1}^{N}(y_n-m)^{2}
        + \lambda_0(\mu_0-m)^{2}
        + (N+\lambda_0)s^{2}
      \Bigr],
  $$

  hence  

  $$
  \boxed{\,b = b_0
      + \dfrac12
        \Bigl(
          \sum_{n=1}^{N}(y_n-m)^{2}
          + \lambda_0(\mu_0-m)^{2}
          + (N+\lambda_0)s^{2}
        \Bigr)\,}.
  $$

---

Therefore  

$$
q^{*}(\tau) \;=\; \operatorname{Gam}\!\bigl(\tau \,\big|\, a,\; b\bigr),
$$

with parameters  

$$
a = a_0 + \frac{N+1}{2},
\qquad
b = b_0
    + \frac12\!
      \left(
        \sum_{n=1}^{N}(y_n-m)^{2}
        + \lambda_0(\mu_0-m)^{2}
        + (N+\lambda_0)s^{2}
      \right).
$$

## 5.1 i

Start from the variational free‑energy 

$$
\mathcal F
= \mathbb E_{q(\theta)}\bigl[\log p(\mathbf y,\theta)\bigr]
  \;-\;
  \mathbb E_{q(\theta)}\bigl[\log q(\theta)\bigr],
  \qquad
  \theta=(\mu,\tau),\;
  q(\theta)=q(\mu)\,q(\tau).
$$

Collecting the $\mu$‑ and $\tau$‑dependent terms (and grouping constants) one obtains  

$$

\begin{aligned}
\mathcal F
&= -\frac{1}{2}\,\mathbb E_{\tau}[\tau]
     \!\left(
        2b_0
        + \mathbb E_{\mu}\!\Bigl[
            \sum_{n=1}^{N}(y_n-\mu)^2
            + \lambda_0(\mu-\mu_0)^2
          \Bigr]
      \right) \\[6pt]
&\quad
  +\left(a_0+\frac{N+1}{2}-1\right)\mathbb E_{\tau}[\log\tau]
  \;-\;\frac{N+1}{2}\log 2\pi
  \;-\;\log\Gamma(a_0) \\[6pt]
&\quad
  +\,a_0\log b_0
  \;+\;\tfrac12\log\lambda_0
  \;-\;\mathbb E_{\mu}[\log q(\mu)]
  \;-\;\mathbb E_{\tau}[\log q(\tau)]\;
\end{aligned}
$$

which is exactly equation (14) in the exercise sheet.

# 5.2: VB Implementation

## 5.2 a

In [ ]:
#  Exercise 5.2 (a) — generate N = 10 observations with true μ = τ = 1

using Random, Distributions          
Random.seed!(42)                     

μ_true  = 1.0                         # ground‑truth mean
τ_true  = 1.0                         # ground‑truth precision  (σ = 1 / √τ)
N       = 10                          # number of samples

y = rand(Normal(μ_true, √(1/τ_true)), N)

@show y            
@show mean(y)      # ≈ 1
@show var(y)       # ≈ 1

y = [0.6366425185482225, 1.2517372155742292, 0.685012028831044, 0.6887475986755793, 1.8163067649323272, 1.476738379831878, 0.1404446179383788, -0.4692882055065464, 0.7933868855188073, 0.6892561269162658]
mean(y) = 0.7708983931260187
var(y) = 0.42190561536738286


0.42190561536738286

## 5.2 b

In [ ]:
import Pkg; Pkg.add("SpecialFunctions")
using Statistics, SpecialFunctions

In [11]:
# prior hyper‑parameters
μ₀, λ₀ = 0.0, 3.0
a₀, b₀ = 2.0, 2.0

# initialise the variational factors  q(μ) · q(τ)
m  = μ₀                # start at the prior mean
s2 = 1 / λ₀            # prior variance of μ
a  = a₀                # prior shape
b  = b₀                # prior rate

maxiter, tol = 1_000, 1e-8
N            = length(y)
ȳ            = mean(y)

for iter in 1:maxiter
    τ̄ = a / b                      #  𝔼_q[τ]
    s2_new = 1 / (τ̄ * (N + λ₀))
    m_new  = s2_new * (λ₀ * μ₀ + N * ȳ)
    a_new = a₀ + (N + 1) / 2                     # shape
    sumsq = sum((y .- m_new).^2)
    b_new = b₀ + 0.5 * (sumsq + λ₀ * (m_new - μ₀)^2 + (N + λ₀) * s2_new)
    if maximum(abs, [m_new-m, s2_new-s2, a_new-a, b_new-b]) < tol
        println("converged after ", iter, " iterations")
        m, s2, a, b = m_new, s2_new, a_new, b_new   # store the last update
        break
    end
    m, s2, a, b = m_new, s2_new, a_new, b_new
end

println("\nPosterior parameters:")
println("  q(μ)  = 𝓝(m = $(round(m, digits=4)),  s² = $(round(s2, digits=4)))")
println("  q(τ)  = Γ(a = $(round(a, digits=4)),  b = $(round(b, digits=4)))")
println("  E[τ]  = ", round(a/b, digits=4))

converged after 10 iterations

Posterior parameters:
  q(μ)  = 𝓝(m = 0.4073,  s² = 0.0528)
  q(τ)  = Γ(a = 7.5,  b = 5.1518)
  E[τ]  = 1.4558


## 5.2 c

In [ ]:
μ₀, λ₀ = 0.0, 3.0
a₀, b₀ = 2.0, 2.0

m  = μ₀
s2 = 1 / λ₀             # prior variance of μ
a  = a₀
b  = b₀

maxiter, tol = 1_000, 1e-8
N            = length(y)
ȳ            = mean(y)

F_hist = Float64[]       # keep the ELBO after each iteration

# -------------------- helper: simplified free energy  (Eq 15) ----------------
free_energy(a, b, a₀, b₀, λ₀, s2, N) = -a*log(b)                  +
                                        loggamma(a) - loggamma(a₀) +
                                        a₀*log(b₀) + 0.5*log(λ₀)   +
                                        0.5*log(s2) - N/2*log(2π)  +
                                        0.5

#  VB iteration loop 
for iter in 1:maxiter
    τ̄ = a / b
    s2_new = 1 / (τ̄ * (N + λ₀))
    m_new  = s2_new * (λ₀ * μ₀ + N * ȳ)
    a_new = a₀ + (N + 1) / 2
    sumsq = sum((y .- m_new).^2)
    b_new = b₀ + 0.5 * (sumsq + λ₀ * (m_new - μ₀)^2 + (N + λ₀) * s2_new)
    push!(F_hist, free_energy(a_new, b_new, a₀, b₀, λ₀, s2_new, N))
    if maximum(abs, [m_new-m, s2_new-s2, a_new-a, b_new-b]) < tol
        println("converged after ", iter, " iterations")
        m, s2, a, b = m_new, s2_new, a_new, b_new   # final parameters
        break
    end

    m, s2, a, b = m_new, s2_new, a_new, b_new
end

println("\nPosterior parameters:")
println("  q(μ)  = 𝓝(m = $(round(m, digits=4)),  s² = $(round(s2, digits=4)))")
println("  q(τ)  = Γ(a = $(round(a, digits=4)),  b = $(round(b, digits=4)))")
println("  E[τ]  = ", round(a/b, digits=4))

println("\nFree‑energy trace:")
println(F_hist)

converged after 10 iterations

Posterior parameters:
  q(μ)  = 𝓝(m = 0.4073,  s² = 0.0528)
  q(τ)  = Γ(a = 7.5,  b = 5.1518)
  E[τ]  = 1.4558

Free‑energy trace:
[-12.698057710571256, -13.003833885744799, -12.98239850044055, -12.985069003649947, -12.984737552497974, -12.984778711185367, -12.984773600525992, -12.984774235119549, -12.98477415632177, -12.984774166106128]


## 5.2 d

In [13]:
μ₀, λ₀ = 0.0, 3.0
a₀, b₀ = 2.0, 2.0

N  = length(y)
ȳ = mean(y)

m, s2 = μ₀, 1/λ₀
a, b  = a₀, b₀


maxiter      = 1_000
tolF         = 1e-3      # free‑energy tolerance
F_prev       = -Inf      # sentinel
F_hist       = Float64[] # store ELBO values

free_energy(a, b, a₀, b₀, λ₀, s2, N) = -a*log(b) +
                                        loggamma(a) - loggamma(a₀) +
                                        a₀*log(b₀) + 0.5*log(λ₀)   +
                                        0.5*log(s2) - N/2*log(2π)  +
                                        0.5

for iter in 1:maxiter
    # 1. 𝔼_q[τ]
    τ̄ = a / b

    # 2. update q(μ)  →  𝓝(m, s²)
    s2 = 1 / (τ̄ * (N + λ₀))
    m  = s2 * (λ₀ * μ₀ + N * ȳ)

    # 3. update q(τ)  →  Γ(a, b)
    a  = a₀ + (N + 1)/2
    sumsq = sum((y .- m).^2)
    b  = b₀ + 0.5 * (sumsq + λ₀ * (m - μ₀)^2 + (N + λ₀) * s2)

    # 4. free‑energy
    F = free_energy(a, b, a₀, b₀, λ₀, s2, N)
    push!(F_hist, F)

    # 5. convergence test on ΔF
    if abs(F - F_prev) < tolF
        println("converged after ", iter, " iterations (|ΔF| < ", tolF, ")")
        break
    end
    F_prev = F
end

println("\nPosterior parameters:")
println("  q(μ) = 𝓝(m = $(round(m, digits=4)),  s² = $(round(s2, digits=4)))")
println("  q(τ) = Γ(a = $(round(a, digits=4)),  b = $(round(b, digits=4)))")
println("  E[τ] = ", round(a/b, digits=4))

println("\nFree‑energy trace:")
println(F_hist)

converged after 5 iterations (|ΔF| < 0.001)

Posterior parameters:
  q(μ) = 𝓝(m = 0.4073,  s² = 0.0528)
  q(τ) = Γ(a = 7.5,  b = 5.1518)
  E[τ] = 1.4558

Free‑energy trace:
[-12.698057710571256, -13.003833885744799, -12.98239850044055, -12.985069003649947, -12.984737552497974]


## 5.1 e

In [32]:
free_energy(a, b, a0, b0, λ0, s2, N) =
      -a * log(b) + loggamma(a) - loggamma(a0) +
       a0 * log(b0) + 0.5 * log(λ0) +
       0.5 * log(s2) - N/2 * log(2π) + 0.5        # (larger = better ELBO)

function vb_univariate_gaussian(y;
        μ0     = 0.0, λ0 = 3.0,        # prior hyper‑parameters
        a0     = 2.0, b0 = 2.0,
        m_init = μ0,                   # initial variational params
        s2_init = 1/λ0,
        a_init = a0,
        b_init = b0,
        tolF   = 1e-3, maxiter = 1000)

    N, ȳ = length(y), mean(y)
    m, s2 = m_init, s2_init
    a, b  = a_init, b_init

    F_prev = -Inf         # previous iteration’s ELBO
    F_curr = free_energy(a, b, a0, b0, λ0, s2, N)   # define once *outside* loop

    for iter in 1:maxiter
        # 1) E_q[τ]
        τ̄ = a / b

        # 2) update q(μ)
        s2 = 1 / (τ̄ * (N + λ0))
        m  = s2 * (λ0 * μ0 + N * ȳ)

        # 3) update q(τ)
        a  = a0 + (N + 1) / 2
        sumsq = sum((y .- m).^2)
        b  = b0 + 0.5 * (sumsq + λ0 * (m - μ0)^2 + (N + λ0) * s2)

        # 4) free energy
        F_curr = free_energy(a, b, a0, b0, λ0, s2, N)

        # 5) convergence test
        abs(F_curr - F_prev) < tolF && break
        F_prev = F_curr
    end

    return (F = F_curr, m = m, s2 = s2, a = a, b = b)
end


#  Run once from the *prior* initial point, then five random starts
using Random, Statistics, SpecialFunctions, Printf

# data & prior already defined y, μ₀, λ₀, a₀, b₀
N, ȳ = length(y), mean(y)

rng = MersenneTwister(2024)
rand_init() = (
    m  = randn(rng),
    s2 = rand(rng, Exponential()),          
    a  = rand(rng, Exponential()) + 1e-3,   
    b  = rand(rng, Exponential()) + 1e-3    
)

results = NamedTuple[]

# (1) prior initialisation
push!(results,
      merge(vb_univariate_gaussian(y;
                                   μ0=μ₀, λ0=λ₀, a0=a₀, b0=b₀,
                                   m_init=μ₀, s2_init=1/λ₀,
                                   a_init=a₀, b_init=b₀),
            (label="prior",)))

# (2) five random initialisations
for k in 1:5
    ini = rand_init()
    res = vb_univariate_gaussian(y;
             μ0=μ₀, λ0=λ₀, a0=a₀, b0=b₀,
             m_init=ini.m,  s2_init=ini.s2,
             a_init=ini.a,  b_init=ini.b)
    push!(results, merge(res, (label="rand$k",)))
end

# sort by ELBO and display
results_sorted = sort(results; by = r -> r.F, rev = true)

println("Summary (sorted by final ELBO 𝓕):")
println(" ──────────────────────────────────────────────────────────────────────────────")
@printf("  %-6s │  𝓕        │     m       s²       a        b\n", "run")
println(" ──────────────────────────────────────────────────────────────────────────────")
for r in results_sorted
    @printf("  %-6s │ %9.4f │ %8.4f %8.4f %8.4f %8.4f\n",
            r.label, r.F, r.m, r.s2, r.a, r.b)
end

Summary (sorted by final ELBO 𝓕):
 ──────────────────────────────────────────────────────────────────────────────
  run    │  𝓕        │     m       s²       a        b
 ──────────────────────────────────────────────────────────────────────────────
  rand3  │  -12.9847 │   0.4073   0.0528   7.5000   5.1518
  rand4  │  -12.9847 │   0.4073   0.0528   7.5000   5.1518
  rand1  │  -12.9847 │   0.4073   0.0528   7.5000   5.1518
  rand5  │  -12.9847 │   0.4073   0.0528   7.5000   5.1518
  prior  │  -12.9847 │   0.4073   0.0528   7.5000   5.1518
  rand2  │  -12.9848 │   0.4073   0.0528   7.5000   5.1518


All initialisations converge to the same solution, so we report the prior‑initialised run: $q(\mu)=\mathcal N(0.4073,\,0.0528),\; q(\tau)=\operatorname{Gam}(7.5,\,5.1518),\; \mathcal F\approx -12.9847$.

## 5.2 f

Because in the normal‑gamma prior the variance of $\mu$ after integrating out $\tau$ is

$$
\operatorname{Var}[\mu]
\;=\; \mathbb E_{\tau}\!\bigl[\frac{1}{\lambda_0\tau}\bigr]
\;=\; \frac{b_0}{a_0\,\lambda_0},
$$

where we used $\mathbb E[\tau]=a_0/b_0$.
Initializing $s^2$ to this value makes the mean‑field factor $q(\mu)$
consistent with the marginal prior and places the very first ELBO exactly at the
prior value.  Starting with $s^2=1/\lambda_0$ would instead correspond to
conditioning on a fixed $\tau=1$, which is incompatible with the prior and
forces an immediate jump in the first update step.